### PR TITLE
Update search to add in object store data for backwards compatibility

### LINKED ISSUE #709: `object_store` key not present in metadata of older object stores
The new version `0.6.0` relies on `object_store` being in the metadata of the Datasource, this isn't the case.
We need a function to migtate the metadata stored.

### PR DESCRIPTION
**Pull request by gareth-j**
* **Summary of changes** (Bug fix, feature, docs update, ...)

Adds in a check for the `object_store` key in the `Datasource` metadata in the search function, if it does exist it's checked, otherwise it's added.

* **Please check if the PR fulfills these requirements**

- [x] Closes #709 
- [x] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing) if fixing a bug or adding a new feature
- [ ] All code checks passing - `black --line-length 110` run over code, `mypy` and `flake8` not showing any errors
- [ ] Documentation and tutorials updated/added
- [x] Added an entry in the latest `CHANGELOG.md` file if fixing a bug or adding a new feature
- [ ] Added any new requirements to `requirements.txt` and `recipes/meta.yaml`

### CODE REVIEW

File: openghg/retrieve/_search.py [hunk 1]
```diff
@@ -605,11 +605,20 @@ def _base_search(**kwargs: Any) -> SearchResults:
             logger.exception(msg=error_msg)
             raise ValueError(error_msg)
 
-        # Here we create a dictionary of the metadata keyed by the Datasource UUID
-        # we'll create a pandas DataFrame out of this in the SearchResult object
-        # for better printing / searching within a notebook
+        # Due to versions of OpenGHG < 0.6.0 not storing the object store
+        # in the metadata we nede to pass that information in here for backwards
```
Comment (rt17603) [inline openghg/retrieve/_search.py:609]:
```suggestion
        # in the metadata we need to pass that information in here for backwards
```

File: openghg/retrieve/_search.py [hunk 2]
```diff
@@ -605,11 +605,20 @@ def _base_search(**kwargs: Any) -> SearchResults:
...
+        # compatibility.
         metadata = {r["uuid"]: r for r in metastore_records}
 
+        updated_metadata = {}
+        store_data = {"object_store": bucket}
```
Comment (rt17603) [inline openghg/retrieve/_search.py:614]: Is this adding in the "object_store" key for every object store this interacts with then? I'm just thinking if the "object_store" key is already present we wouldn't want this to be silently replaced here without us realising so may want to add a check if this is present before replacing it.

**Review by rt17603 — CHANGES_REQUESTED** [review]
Question around behaviour when "object_store" key is already present in the metadata - is this always being replaced at the moment? If so, we should update this to be a little more cautious and not overwrite existing keys.

File: openghg/retrieve/_search.py (hunk 2 above)
Comment (gareth-j) [reply openghg/retrieve/_search.py:614]: Yeah I'm going to add a check in there and make sure the values don't differ.

File: openghg/retrieve/_search.py [hunk 3]
```diff
@@ -605,11 +605,29 @@ def _base_search(**kwargs: Any) -> SearchResults:
...
-        # Here we create a dictionary of the metadata keyed by the Datasource UUID
-        # we'll create a pandas DataFrame out of this in the SearchResult object
-        # for better printing / searching within a notebook
+        # Due to versions of OpenGHG < 0.6.0 not storing the object store
+        # in the metadata we need to pass that information in here for backwards
+        # compatibility.
         metadata = {r["uuid"]: r for r in metastore_records}
 
+        updated_metadata = {}
+        store_data = {"object_store": bucket}
```
Comment (rt17603) [inline openghg/retrieve/_search.py:614]: How is bucket being defined (and how do we want it to be defined) - at the moment it looks like it's just the last item in the loop on L589?

File: openghg/retrieve/_search.py [hunk 4]
```diff
@@ -605,11 +605,29 @@ def _base_search(**kwargs: Any) -> SearchResults:
...
+
+        for uid, _metadata in metadata.items():
+            if "object_store" in _metadata:
+                ds_bucket = _metadata["object_store"]
+                if ds_bucket != bucket:
+                    raise ObjectStoreError(
+                        f"Mismatch between Datasource object_store: {ds_bucket} and "
```
Comment (rt17603) [inline openghg/retrieve/_search.py:621]: I'm a little cautious about adding in an error here that users won't be able to do anything about - any ideas why this error might come in and what advice we could give to fix this?

File: openghg/retrieve/_search.py (hunk 3 above)
Comment (gareth-j) [reply openghg/retrieve/_search.py:614]: Yeah we iterate over the dictionary `readable_buckets` that's a store_name: store_path pair.

Comment (rt17603) [reply openghg/retrieve/_search.py:614]: But won't `bucket` just be set to the last value taken from `readable_buckets` from the loop? Is that what is expected?

Comment (rt17603) [reply openghg/retrieve/_search.py:614]: Ok - right this part is in the loop - I missed the nesting here.

**Pull request closed without merging**